src/compiler.ts

```diff
@@ -6405,12 +6453,58 @@ export class Compiler extends DiagnosticEmitter {
       case ElementKind.LOCAL: {
         let local = <Local>target;
         signature = local.type.signatureReference;
+        if(local.type.is(TypeFlags.CLOSURE)) {
+          indexArg = module.block(null, [
+            this.injectClosedLocals(
+              this.module.local_get(local.index, this.options.nativeSizeType),
+              local.type
+            ),
+            module.global_set(
+              BuiltinNames.global_closure,
+              this.module.local_get(local.index, this.options.nativeSizeType)
+            ),
+            module.load(
+              local.type.byteSize,
+              local.type.is(TypeFlags.SIGNED),
+              this.module.local_get(local.index, this.options.nativeSizeType),
+              this.options.nativeSizeType,
+              0
+            )
+          ], this.options.nativeSizeType);
+          break;
+        }
         if (signature) {
           if (local.is(CommonFlags.INLINED)) {
             indexArg = module.i32(i64_low(local.constantIntegerValue));
           } else {
             indexArg = module.local_get(local.index, NativeType.I32);
           }
+          // If the 16 is divisible by 16, then it must be a closure, so we call it as such
+          indexArg = module.if(
```

Comment:
Outside of the local scope, we only have function pointers, we don't know ahead of time whether or not it'll be a closure. So instead we check to see if it's divisible by 16, and if it is assume its a pointer. We don't know what locals this closure uses, but we don't need to since the local scope it was referencing is inaccessible to us here anyway. We just set the global closure and assume the function will know how to read from that.

Reply:
This has been changed, now we just check to see if the function reference has its MSB set, in which case we assume its a closure.